File: web/live.js
import {CURRENT_LIMITS,FORWARD,REVERSE,SENSORY,meanTrace,workerBase,validateLiveSnapshot,validateTokenMarket} from './report.js';
const $=id=>document.getElementById(id);
let connection=null,pollTimer=null,marketTimer=null,publishedTimer=null,busy=false,lastRoot=null,recordedSnapshot=false,publishedRecording=null,publishedIndex=0;
let marketOnline=false,marketAttempted=false,observerStatus='idle';
const terminal=new Set(['idle','stopped','failed']);
const observerActive=()=>!terminal.has(observerStatus)&&observerStatus!=='disconnected';
function status(text){$('live-status').textContent=text;}
function paintState(){
  $('live-state').textContent=observerActive()?observerStatus.toUpperCase().replaceAll('-',' '):recordedSnapshot?`RECORDED ${publishedRecording?.frames.length??0} TICKS`:marketOnline?'TOKEN LIVE':marketAttempted?'FEED RETRYING':'CONNECTING';
}
function clock(seconds){return `T + ${String(Math.floor(seconds/60)).padStart(2,'0')}:${String(seconds%60).padStart(2,'0')}`;}
function stopPublishedPlayback(){clearInterval(publishedTimer);publishedTimer=null;$('live-replay').textContent='▶ Replay observer run';}
async function request(path,method='GET'){
  const response=await fetch(connection.base+path,{method,headers:connection.token?{Authorization:`Bearer ${connection.token}`}:{},redirect:'error',signal:AbortSignal.timeout(15000)});
  if(!response.ok)throw new Error(response.status===401?'Worker token rejected.':response.status===429?'Worker is already in use.':'Cannot reach the neural worker. Check its URL and access token below.');
  return response.json();
}
function trace(frame){
  const svg=$('live-chart'),series=[meanTrace(frame,FORWARD),meanTrace(frame,REVERSE)];
  const values=series.flat(),low=Math.min(...values)-.01,high=Math.max(...values)+.01;
  const ns='http://www.w3.org/2000/svg';svg.replaceChildren();
  series.forEach((v,j)=>{const line=document.createElementNS(ns,'polyline');line.setAttribute('points',v.map((y,i)=>`${10+i/(v.length-1)*530},${10+(high-y)/(high-low)*145}`).join(' '));line.setAttribute('fill','none');line.setAttribute('stroke',j?'#f6b396':'#a5f4cb');line.setAttribute('stroke-width','2');svg.append(line);});
  const label=document.createElementNS(ns,'text');label.setAttribute('x','10');label.setAttribute('y','176');label.setAttribute('class','chart-label');label.textContent=`${low.toFixed(2)} to ${high.toFixed(2)} mV · latest neural episode`;svg.append(label);
}
function renderMarket(value){
  const r=validateTokenMarket(value),m=r.recent;
  marketOnline=true;marketAttempted=true;paintState();
  $('live-count').textContent=m.events.toLocaleString();
  $('live-sides').textContent=`${m.buy_events} / ${m.sell_events}`;
  $('live-price').textContent=m.price_quote===null?'Unavailable':m.price_quote.toPrecision(5);
  const age=m.latest_trade_at===null?null:Math.max(0,Math.floor(Date.now()/1000-m.latest_trade_at));
  $('live-coverage').textContent=`Pons indexer · recent indexed window · latest indexed trade ${age===null?'unavailable':age+'s ago'}. Amounts are in GOOGL; receipts, liquidity and executable price impact are unverified.`;
  if(!observerActive())status(recordedSnapshot?'Live WORMBRAIN token data connected. The neural panel shows the published continuous observer run; start a worker for fresh neural ticks.':'Live WORMBRAIN token data connected. The optional Python worker below runs the c302 neural observer.');
}
async function pollMarket(){
  try{
    const response=await fetch('/api/token',{headers:{Accept:'application/json'},redirect:'error',signal:AbortSignal.timeout(15000)});
    if(!response.ok)throw new Error('Token feed unavailable');
    renderMarket(await response.json());
  }catch{
    marketOnline=false;marketAttempted=true;paintState();
    if(!observerActive())status('The WORMBRAIN token feed is temporarily unavailable. Retrying automatically.');
  }finally{marketTimer=setTimeout(pollMarket,15000);}
}
function render(value,{recorded=false}={}){
  const r=validateLiveSnapshot(value),frame=recorded?r.frames[publishedIndex]:r.frames.at(-1);observerStatus=r.status;paintState();
  $('live-stop').disabled=terminal.has(r.status);$('live-export').disabled=!frame;
  if(!frame){status(r.status==='preparing-brain'?'Preparing the complete generated network. Export and native compilation can take a few minutes.':r.error||'Waiting for the first real transaction observation.');return;}
  const m=frame.market;
  status(recorded?'Published continuous token observer run. Absolute wall-clock fields were removed; relative neural and paper-risk timelines remain.':r.error||(r.status==='stopped'?'Session stopped. Live token data remains connected.':`Observed ${r.total_stimulated_events} fresh trades for stimulation. ${m.feed_status==='observed-trades'?'The brain, plastic gain, decoder, and paper-risk state persist between updates.':'No new fresh trades: external sensory currents are zero.'}`));
  if(!recorded){
    const age=m.latest_trade_at?Math.max(0,Math.floor(Date.now()/1000-m.latest_trade_at)):null;
    $('live-count').textContent=r.total_observed_events.toLocaleString();$('live-sides').textContent=`${m.buy_events??0} / ${m.sell_events??0}`;$('live-price').textContent=m.price_quote?m.price_quote.toPrecision(5):'Unavailable';
    $('live-coverage').textContent=`Pons indexer · ${m.coverage} · ${m.gap_count} polling gaps · latest indexed trade ${age===null?'unavailable':age+'s ago'}. Amounts are in GOOGL; receipts, liquidity and executable price impact are unverified.`;
  }else{
    const tick=frame.paper_tick;
    $('live-scrubber').value=publishedIndex;$('live-frame-label').textContent=`${frame.decoder.state} · ${frame.risk}`;$('live-frame-time').textContent=clock(tick.seconds);$('live-frame-count').textContent=`${publishedIndex+1} / ${r.frames.length}`;
    $('live-frame-summary').textContent=`${m.new_events} new indexed events · ${m.stimulated_events} fresh stimuli · recorded price ${m.price_quote?m.price_quote.toPrecision(5):'unavailable'} GOOGL · ${m.feed_status.replaceAll('-',' ')}`;
  }
  const renderKey=`${r.audit_root}:${frame.event_hash}`;if(lastRoot===renderKey)return;lastRoot=renderKey;
  $('live-forward').textContent=frame.scores.forward.toFixed(3);$('live-reverse').textContent=frame.scores.reverse.toFixed(3);$('live-time').textContent=`${frame.time_ms.toFixed(0)} ms`;
  $('live-decoder').textContent=frame.decoder?.state??'LEGACY';
  $('live-rim-rib').textContent=frame.scores.vector?.rim==null?'NOT RECORDED':`${frame.scores.vector.rim.toFixed(3)} / ${frame.scores.vector.rib.toFixed(3)}`;
  $('live-plasticity').textContent=frame.plasticity?`${frame.plasticity.applied_gain.toFixed(3)} → ${frame.plasticity.next_gain.toFixed(3)}`:'NOT ACTIVE';
  const feedback=frame.next_feedback_pa;$('live-feedback').textContent=feedback?`${Math.max(feedback.PVCL,feedback.PVCR).toFixed(2)} / ${feedback.DVA.toFixed(2)} pA`:'NOT ACTIVE';
  $('live-risk').textContent=frame.risk?`${frame.risk} · ${frame.execution}`:'OBSERVER ONLY';
  $('live-reasons').textContent=frame.reasons?.length?frame.reasons.join(' · '):'NONE';
  $('live-model-proof').textContent=r.model?`${r.model.neurons} neurons · ${r.model.projections} projections · ${(r.model.reader??'unknown reader').replace('cect.readers.','')} · model ${r.model.network_sha256?.slice(0,12)??'unpublished'} · reader ${r.model.reader_cache_sha256?.slice(0,12)??'legacy'}`:'Model identity unavailable.';
  $('live-sensory').replaceChildren(...SENSORY.map(name=>{const row=document.createElement('div');row.className='sensory-row';const label=document.createElement('span');label.textContent=name;const value=document.createElement('span');value.textContent=frame.stimulus_pa[name].toFixed(2)+' pA';const bar=document.createElement('progress');bar.max=CURRENT_LIMITS[name];bar.value=frame.stimulus_pa[name];bar.setAttribute('aria-label',name+' live current');row.append(label,value,bar);return row;}));
  trace(frame);
}
async function poll(){
  try{const value=await request('/api/live');render(value);if(terminal.has(value.status)){$('live-start').disabled=false;return;}}
  catch(error){observerStatus='disconnected';paintState();status(marketOnline?`Live token data is connected. ${error.message}`:error.message);$('live-start').disabled=false;return;}
  pollTimer=setTimeout(poll,5000);
}
async function loadPublishedRecording(){
  try{
    const response=await fetch('/data/live-observer.json',{headers:{Accept:'application/json'},redirect:'error',signal:AbortSignal.timeout(15000)});
    if(!response.ok)return;
    publishedRecording=validateLiveSnapshot(await response.json());publishedIndex=publishedRecording.frames.length-1;recordedSnapshot=true;$('live-recording-link').hidden=false;$('live-timeline').hidden=false;$('live-scrubber').max=publishedIndex;render(publishedRecording,{recorded:true});paintState();
  }catch{}
}
$('live-scrubber').addEventListener('input',event=>{if(!publishedRecording)return;stopPublishedPlayback();publishedIndex=Number(event.target.value);lastRoot=null;render(publishedRecording,{recorded:true});});
$('live-replay').addEventListener('click',()=>{if(!publishedRecording)return;if(publishedTimer){stopPublishedPlayback();return;}publishedIndex=0;lastRoot=null;render(publishedRecording,{recorded:true});$('live-replay').textContent='Ⅱ Pause observer run';publishedTimer=setInterval(()=>{if(++publishedIndex>=publishedRecording.frames.length){publishedIndex=publishedRecording.frames.length-1;stopPublishedPlayback();return;}lastRoot=null;render(publishedRecording,{recorded:true});},1800);});
$('live-start').addEventListener('click',async()=>{
  if(busy)return;busy=true;clearTimeout(pollTimer);stopPublishedPlayback();lastRoot=null;
  try{
    connection={base:workerBase($('worker-url').value,location.origin),token:$('worker-token').value};
    const current=validateLiveSnapshot(await request('/api/live'));
    recordedSnapshot=false;$('live-timeline').hidden=true;
    if(terminal.has(current.status)){await request('/api/live/start','POST');observerStatus='preparing-brain';paintState();}
    $('live-start').disabled=true;$('live-stop').disabled=false;await poll();
  }catch(error){observerStatus='disconnected';if(publishedRecording){recordedSnapshot=true;$('live-timeline').hidden=false;lastRoot=null;render(publishedRecording,{recorded:true});}paintState();status(`${marketOnline?'Live token data is connected. ':''}${error.message}${publishedRecording?' The published observer run remains available.':''}`);$('live-start').disabled=false;}finally{busy=false;}
});
$('live-stop').addEventListener('click',async()=>{try{await request('/api/live/stop','POST');status('Stopping the neural session. Live token data remains connected.');}catch(error){status(error.message);}});
$('live-export').addEventListener('click',async()=>{try{const value=recordedSnapshot?validateLiveSnapshot(publishedRecording):validateLiveSnapshot(await request('/api/live?history=true'));const url=URL.createObjectURL(new Blob([JSON.stringify(value,null,2)],{type:'application/json'}));const a=document.createElement('a');a.href=url;a.download='wormbrain-live-observations.json';a.click();setTimeout(()=>URL.revokeObjectURL(url),1000);}catch(error){status(error.message);}});

paintState();
pollMarket();
loadPublishedRecording();

File: src/wormbrain/live.py
"""Bounded live observation session, isolated from the paper replay executor."""
from __future__ import annotations
import copy
import threading
import time

from .core import (MODEL_CONFIG, PLASTICITY, POLICY, PROPRIOCEPTION, PROPRIOCEPTIVE,
                   READOUT, SENSORY, SENSORY_CALIBRATION, DopaminePlasticity, PaperEngine,
                   digest, neural_scores, proprioceptive_feedback)
from .pons import FeedError, LIVE_ENCODING, POLL_SECONDS, SOURCE, TradeFeed, fetch_trades
from .stream import LiveBrain


def sanitize_public_recording(recording: dict) -> dict:
    """Remove wall clocks and rebuild the audit chain for publication."""
    if (recording.get("schema_version") != 2
            or recording.get("kind") != "wormbrain-live-observer"
            or recording.get("trading_mode") != "paper"
            or recording.get("live_execution") is not False):
        raise ValueError("Not a supported live observer recording")
    frames = recording.get("frames")
    if not isinstance(frames, list) or len(frames) < 2:
        raise ValueError("A public continuous recording needs at least two frames")
    public = copy.deepcopy(recording)
    previous = "genesis"
    for frame in public["frames"]:
        frame.pop("event_hash", None)
        frame.pop("previous_hash", None)
        frame.get("market", {}).pop("latest_trade_at", None)
        if set(frame.get("stimulus_pa", {})) != set(SENSORY):
            raise ValueError("Recording has incomplete sensory currents")
        required = set(READOUT)
        if not required.issubset(public.get("recorded_cells", ())) or not required.issubset(frame.get("voltage_mv", {})):
            raise ValueError("Recording has incomplete public voltages")
        frame["previous_hash"] = previous
        previous = digest(frame)
        frame["event_hash"] = previous
    public.update(status="stopped", error=None, audit_root=previous,
                  recording={"kind": "continuous-token-observer",
                             "clock": "relative tick and biological model time",
                             "privacy": "absolute polling and market-event times removed"})
    return public


class LiveMonitor:
    def __init__(self, *, fetcher=fetch_trades, brain_factory=LiveBrain):
        self.fetcher = fetcher
        self.brain_factory = brain_factory
        self.lock = threading.Lock()
        self.stop_event = threading.Event()
        self.thread = None
        self.state = dict(schema_version=2, kind="wormbrain-live-observer", status="idle",
                          source=SOURCE, encoding=LIVE_ENCODING, trading_mode="paper", live_execution=False,
                          policy=POLICY, calibration=SENSORY_CALIBRATION,
                          plasticity=PLASTICITY, proprioception=PROPRIOCEPTION,
                          frames=[], total_observed_events=0, total_stimulated_events=0,
                          audit_root="genesis", error=None)

    @property
    def active(self):
        return bool(self.thread and self.thread.is_alive())

    def snapshot(self, *, history=True):
        with self.lock:
            return copy.deepcopy({**self.state, "frames": self.state["frames"] if history else self.state["frames"][-1:]})

    def start(self, *, seconds=3600):
        if type(seconds) is not int or not 1 <= seconds <= 86400:
            raise ValueError("Session must be 1–86400 seconds")
        if self.active:
            raise ValueError("Live observation is already running")
        self.stop_event.clear()
        with self.lock:
            self.state.update(status="preparing-brain", frames=[], total_observed_events=0,
                              total_stimulated_events=0, audit_root="genesis", error=None,
                              session_elapsed_seconds=0, requested_seconds=seconds)
        self.thread = threading.Thread(target=self._run, args=(seconds,), daemon=True)
        self.thread.start()

    def stop(self):
        self.stop_event.set()
        with self.lock:
            if self.active:
                self.state["status"] = "stopping"

    def _run(self, seconds):
        feed = TradeFeed()
        engine = PaperEngine()
        plasticity = DopaminePlasticity()
        feedback = {name: 0.0 for name in PROPRIOCEPTIVE}
        fallback_price = 1.0
        tick_index = 0
        previous = "genesis"
        try:
            with self.brain_factory() as brain:
                deadline = time.monotonic() + seconds
                with self.lock:
                    self.state.update(status="connecting", model=brain.model, config=MODEL_CONFIG,
                                      null_mv=brain.null_mv, recorded_cells=brain.recorded_cells,
                                      plasticity_targets=brain.plasticity_targets)
                while not self.stop_event.is_set() and time.monotonic() < deadline:
                    begin = time.monotonic()
                    error = None
                    try:
                        observations = self.fetcher()
                        summary, currents = feed.ingest(observations)
                    except FeedError as exc:
                        summary = dict(observed_at=int(time.time()), new_events=0, stimulated_events=0,
                                       total_events=feed.total_events, feed_status="unavailable", coverage="unknown",
                                       latest_trade_at=feed.latest_timestamp, gap_count=feed.gap_count)
                        currents = {n: 0.0 for n in SENSORY}
                        error = str(exc)
                    if self.stop_event.is_set():
                        break
                    # Even during an outage the same network advances with zero external current.
                    applied_feedback = feedback
                    neural = brain.step(currents, feedback_pa=applied_feedback, plasticity_gain=plasticity.gain)
                    scores = neural_scores(neural["voltage_mv"], brain.null_mv)
                    if summary.get("price_quote"):
                        fallback_price = summary["price_quote"]
                    paper_tick = dict(tick=tick_index, seconds=tick_index * POLL_SECONDS,
                                      price=fallback_price, flow=summary.get("flow", 0.0),
                                      liquidity=0.0, price_impact_bps=0.0,
                                      fresh=bool(summary.get("stimulated_events")), eligible=True,
                                      liquidity_verified=False, price_impact_verified=False)
                    outcome = engine.step(paper_tick, scores)
                    plasticity_frame = plasticity.update(currents)
                    feedback = proprioceptive_feedback(outcome)
                    summary.pop("observed_at", None)
                    body = dict(market=summary, stimulus_pa=currents, time_ms=neural["time_ms"],
                                voltage_mv=neural["voltage_mv"], scores=scores, traces=neural["traces"],
                                paper_tick=paper_tick, plasticity=plasticity_frame,
                                feedback_pa=applied_feedback, next_feedback_pa=feedback,
                                previous_hash=previous, **outcome)
                    previous = digest(body)
                    row = dict(body, event_hash=previous)
                    with self.lock:
                        self.state["frames"] = (self.state["frames"] + [row])[-240:]
                        self.state.update(status="source-unavailable" if error else "observing", error=error,
                                          audit_root=previous, total_observed_events=feed.total_events,
                                          session_elapsed_seconds=tick_index * POLL_SECONDS)
                        self.state["total_stimulated_events"] += summary["stimulated_events"]
                    tick_index += 1
                    remaining = min(POLL_SECONDS - (time.monotonic() - begin), deadline - time.monotonic())
                    if remaining > 0:
                        self.stop_event.wait(remaining)
        except Exception:
            with self.lock:
                self.state.update(status="failed", error="Live model failed. Check Java and the native compiler locally.")
            return
        with self.lock:
            self.state["status"] = "stopped"


monitor = LiveMonitor()
